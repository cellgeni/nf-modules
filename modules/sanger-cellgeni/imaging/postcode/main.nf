process IMAGING_POSTCODE {
    tag "${meta.id}"

    container "quay.io/cellgeni/postcode:0.2.0"

    input:
    tuple val(meta), file(profile), file(spot_loc), file(starfish_codebook), file(readouts), val(R)

    output:
    tuple val(meta), path("${out_name}"), emit: decoded_peaks
    tuple val(meta), path("${prefix}_decode_out_parameters.pickle"), optional: true
    path "versions.yml", emit: versions

    script:
    prefix = meta.id ?: "none"
    out_name = "${prefix}_decoded_spots.csv"
    readouts = readouts ? "--readouts ${readouts}" : ""
    def args = task.ext.args ?: ""
    """
    decode.py run \
        --spot_profile_p ${profile} \
        --spot_locations_p ${spot_loc} \
        --codebook_p ${starfish_codebook} \
        --out_name ${out_name} \
        --R ${R} \
        ${readouts}  \
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bioinfotongli: \$(decode.py version)
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    out_name = "${prefix}_decoded_spots.csv"
    """
    touch ${out_name}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bioinfotongli: \$(decode.py version)
    END_VERSIONS
    """
}
