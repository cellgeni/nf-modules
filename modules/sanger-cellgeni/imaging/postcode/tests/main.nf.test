// nf-core modules test imaging/postcode
nextflow_process {

    name "Test Process IMAGING_POSTCODE"
    script "../main.nf"
    process "IMAGING_POSTCODE"

    tag "modules"
    tag "modules_sanger-cellgeni"
    tag "imaging"
    tag "imaging/postcode"

    test("merfish - npy - 1") {

        when {
            params.module_args = '--train_params {up_prc_to_remove:99.95}'
            process {
                """
                input[0] = [
                    [ id:'merfish_1'],
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_reformatted_profile.npy', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_starfish_codebook.json', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_barcodes_0123_str.txt', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_1.csv', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("merfish - npy - 5000") {

        when {
            params.module_args = '--train_params {up_prc_to_remove:99.95}'
            process {
                """
                input[0] = [
                    [ id:'merfish_5000'],
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_5000_reformatted_profile.npy', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_5000_starfish_codebook.json', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_5000_barcodes_0123_str.txt', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_5000.csv', checkIfExists: true),
                ] 
                """
            }
        }

        then {
            assertAll(
                // { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("merfish - npy - 100000") {

        when {
            params.module_args = '--train_params {up_prc_to_remove:99.95}'
            process {
                """
                input[0] = [
                    [ id:'merfish_100000'],
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_100000_reformatted_profile.npy', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_100000_starfish_codebook.json', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_100000_barcodes_0123_str.txt', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_100000.csv', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("stub") {

        options "-stub"
        
        when {
            params.module_args = '--train_params {up_prc_to_remove:99.95}'
            process {
                """
                input[0] = [
                    [ id:'merfish_test'],
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_reformatted_profile.npy', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_starfish_codebook.json', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/for_decoding/merfish_1_barcodes_0123_str.txt', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_1.csv', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
