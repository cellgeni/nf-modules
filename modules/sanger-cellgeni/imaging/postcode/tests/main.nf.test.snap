{
    "merfish - npy - 1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "merfish_1"
                        },
                        "merfish_1_decoded_spots.csv:md5,2fe3567064eb803bb46e950a76a5ec62"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ],
                "decoded_peaks": [
                    [
                        {
                            "id": "merfish_1"
                        },
                        "merfish_1_decoded_spots.csv:md5,2fe3567064eb803bb46e950a76a5ec62"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-14T19:51:48.628437149"
    },
    "merfish - csv": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "decoded_peaks": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-11T15:19:07.13337354"
    },
    "merfish - npy - 100000": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "merfish_100000"
                        },
                        "merfish_100000_decoded_spots.csv:md5,771ccb4c887ef4ec6f4232f10e69125d"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ],
                "decoded_peaks": [
                    [
                        {
                            "id": "merfish_100000"
                        },
                        "merfish_100000_decoded_spots.csv:md5,771ccb4c887ef4ec6f4232f10e69125d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-14T21:38:55.953642885"
    },
    "merfish - npy - 5000": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "merfish_5000"
                        },
                        "merfish_5000_decoded_spots.csv:md5,c3f53ca490b62ec46d85591c4770cc01"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ],
                "decoded_peaks": [
                    [
                        {
                            "id": "merfish_5000"
                        },
                        "merfish_5000_decoded_spots.csv:md5,c3f53ca490b62ec46d85591c4770cc01"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-16T14:46:12.869234749"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "merfish_test"
                        },
                        "merfish_test_decoded_spots.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ],
                "decoded_peaks": [
                    [
                        {
                            "id": "merfish_test"
                        },
                        "merfish_test_decoded_spots.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fd95950f4ce1bcca31743bd79265a816"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-14T19:21:38.159156009"
    }
}